torchdata/datapipes/iter/load/fsspec.py

```diff
@@ -49,37 +49,45 @@ class FSSpecFileListerIterDataPipe(IterDataPipe[str]):
 
     def __init__(
         self,
-        root: str,
+        root: Union[str, Sequence[str], IterDataPipe],
         masks: Union[str, List[str]] = "",
     ) -> None:
         _assert_fsspec()
 
-        self.root: str = root
+        if isinstance(root, str):
+            root = [
+                root,
+            ]
+        if not isinstance(root, IterDataPipe):
+            self.datapipe: Union[IterDataPipe, IterableWrapper] = IterableWrapper(root)
+        else:
+            self.datapipe = root
         self.masks = masks
 
     def __iter__(self) -> Iterator[str]:
-        fs, path = fsspec.core.url_to_fs(self.root)
-        is_local = fs.protocol == "file" or not self.root.startswith(fs.protocol)
-        if fs.isfile(path):
-            yield self.root
-        else:
-            for file_name in fs.ls(path):
-                if not match_masks(file_name, self.masks):
-                    continue
-
-                # ensure the file name has the full fsspec protocol path
-                if file_name.startswith(fs.protocol):
-                    yield file_name
-                else:
-                    if is_local:
-                        abs_path = os.path.join(path, file_name)
-                    else:
-                        abs_path = posixpath.join(path, file_name)
-
-                    if self.root.startswith(fs.protocol):
-                        yield fs.protocol + "://" + abs_path
+        for root in self.datapipe:
+            fs, path = fsspec.core.url_to_fs(root)
+            is_local = fs.protocol == "file" or not root.startswith(fs.protocol)
+            if fs.isfile(path):
+                yield root
+            else:
+                for file_name in fs.ls(path):
+                    if not match_masks(file_name, self.masks):
+                        continue
+
+                    # ensure the file name has the full fsspec protocol path
+                    if file_name.startswith(fs.protocol):
+                        yield file_name
                     else:
-                        yield abs_path
+                        if is_local:
+                            abs_path = os.path.join(path, file_name)
+                        else:
+                            abs_path = posixpath.join(path, file_name)
+
+                        if root.startswith(fs.protocol):
+                            yield fs.protocol + "://" + abs_path
```

Comment:
The logic here might need to be changed as well.

Reply:
In what way should this be changed?

Reply:
I believe @ejguan is referring to his other comment here:
https://github.com/pytorch/data/pull/383#discussion_r866189401

It will be great if you can solve #348 as well in this PR, and when you do, the logic here will have to change as well.

Reply:
```suggestion
                        if isinstance(fs.protocol, list):
                            for protocol in fs.protocol:
                                if root.startswith(protocol):
                                   yield protocol + "://" + abs_path
                                   break
                        elif root.startswith(fs.protocol):
                            yield fs.protocol + "://" + abs_path
```